src/main/java/com/atlassian/db/replica/internal/state/ConnectionState.java

```diff
@@ -53,9 +54,16 @@ public ConnectionState(
             @Override
             public Connection create() throws SQLException {
                 if (connectionProvider.isReplicaAvailable()) {
-                    final Connection replicaConnection = connectionProvider.getReplicaConnection();
-                    parameters.initialize(replicaConnection);
-                    return replicaConnection;
+                    try {
+                        final Connection replicaConnection = connectionProvider.getReplicaConnection();
+                        parameters.initialize(replicaConnection);
+                        return replicaConnection;
+                    } catch (SQLException e) {
+                        return replicaFailureStrategy.onFailure(
```

Comment:
have you thought about more sophisticated mechanism aka CirucitBreaker configuration:
How often it will be run ? on every failed call ? fallback only after 5 failed calls, happened on x concurrent threads, and lasting x minutes? 
Should it retry  ?
I've just listed my ideas, doesn't insist to implement all of them, more want to understood your approach for them.

Reply:
> have you thought about more sophisticated mechanism

Yes. That's why it's an SPI.